File: client/src/test/scala/org/zalando/nakadi/client/scala/model/ModelFactory.scala
package org.zalando.nakadi.client.scala.model

import java.util.UUID
import com.fasterxml.jackson.annotation.JsonProperty
import scala.util.Random

case class BusinessEventImpl(metadata: Option[EventMetadata]) extends BusinessEvent
case class SimpleEvent(id: String, name: String) extends Event

object ModelFactory {
  def randomUUID(): String = UUID.randomUUID().toString();
  def randomInt(): Integer = Random.nextInt()
  def randomListOfStrings(nrOfItems: Int): Seq[String] = for (i <- 1 to nrOfItems) yield randomUUID()
  def randomMapOfStrings(nrOfItems: Int): Map[String, Any] = (for (i <- 1 to nrOfItems) yield (s"itemNr $i" -> randomMapOfStrings(nrOfItems - 1))).toMap

  def newBatchItemResponse(): BatchItemResponse = {
    val detail = Option("detail")
    val step = Option(BatchItemStep.ENRICHING)
    val publishingStatus = BatchItemPublishingStatus.SUBMITTED
    val eid = Option(randomUUID())
    new BatchItemResponse(eid, publishingStatus, step, detail)
  }
  def newBusinessEvent() = new BusinessEventImpl(Option(newEventMetadata))
  def newEventMetadata(): EventMetadata = new EventMetadata(randomUUID(), Option("eventType"), "occurredAt", Option("receivedAt"), List("parentEids"), Option(randomUUID()), Option("partition"))
  def newCursor(): Cursor = new Cursor("partition", "offset")
  def newSimpleEvent(): SimpleEvent = SimpleEvent(randomUUID(), randomUUID())
  def newDataChangeEvent[T <: Event](in: T): DataChangeEvent[T] = DataChangeEvent(in, randomUUID(), DataOperation.CREATE, Option(newEventMetadata()))
  def newEventStreamBatch[T <: Event](events: List[T], cursor: Cursor): EventStreamBatch[T] = EventStreamBatch(cursor, Option(events))
  def newEventTypeSchema(): EventTypeSchema = {
    val schema = "{'propeties':{'key':'value'}}".replaceAll("'", "\"");
    EventTypeSchema(SchemaType.JSON, schema);
  }
  def newEventTypeStatistics(): EventTypeStatistics = EventTypeStatistics(randomInt(), randomInt(), randomInt(), randomInt())
  def newMetrics(): Metrics = {
    val depth = 5
    val gauges: Map[String, Any] = Map[String, Any](
      "normalString" -> randomUUID(),
      "SimpleEvent" -> newSimpleEvent(),
      "List" -> randomListOfStrings(depth),
      "MapOfMaps" -> randomMapOfStrings(depth))
    Metrics(randomUUID(), gauges)
  }
  def newPartition(): Partition = Partition(randomUUID(), randomUUID(), randomUUID())
  def newProblem(): Problem = Problem(randomUUID(), randomUUID(), randomInt(), Option(randomUUID()), Option(randomUUID()))

  def newPartitionStrategy(): PartitionStrategy.Value = {
    val values = PartitionStrategy.values.toArray
    values(Random.nextInt(values.length))
  }

  def newEventEnrichmentStrategy(): EventEnrichmentStrategy.Value = {
    val values = EventEnrichmentStrategy.values.toArray
    values(Random.nextInt(values.length))
  }

  def newEventTypeCategory(): EventTypeCategory.Value = {
    val values = EventTypeCategory.values.toArray
    values(Random.nextInt(values.length))
  }

  def newEventType(): EventType = new EventType(randomUUID(), //name
    randomUUID(), //owningApplication
    newEventTypeCategory(), //category
    List(newEventEnrichmentStrategy(), newEventEnrichmentStrategy()), //enrichmentStrategies
    Option(newPartitionStrategy()), //
    newEventTypeSchema(), //
    List(randomUUID(),
      randomUUID()), //dataKeyFields
    List(randomUUID(),
      randomUUID()), //dataKeyFields
    Option(newEventTypeStatistics))
}

File: client/src/test/java/org/zalando/nakadi/client/java/model/ModelFactory.java
package org.zalando.nakadi.client.java.model;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Random;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.zalando.nakadi.client.java.ClientError;
import org.zalando.nakadi.client.java.enumerator.BatchItemPublishingStatus;
import org.zalando.nakadi.client.java.enumerator.BatchItemStep;
import org.zalando.nakadi.client.java.enumerator.DataOperation;
import org.zalando.nakadi.client.java.enumerator.EventEnrichmentStrategy;
import org.zalando.nakadi.client.java.enumerator.EventTypeCategory;
import org.zalando.nakadi.client.java.enumerator.PartitionStrategy;
import org.zalando.nakadi.client.java.enumerator.SchemaType;
import org.zalando.nakadi.client.java.model.BatchItemResponse;
import org.zalando.nakadi.client.java.model.BusinessEvent;
import org.zalando.nakadi.client.java.model.Cursor;
import org.zalando.nakadi.client.java.model.DataChangeEvent;
import org.zalando.nakadi.client.java.model.Event;
import org.zalando.nakadi.client.java.model.EventMetadata;
import org.zalando.nakadi.client.java.model.EventStreamBatch;
import org.zalando.nakadi.client.java.model.EventType;
import org.zalando.nakadi.client.java.model.EventTypeSchema;
import org.zalando.nakadi.client.java.model.EventTypeStatistics;
import org.zalando.nakadi.client.java.model.Metrics;
import org.zalando.nakadi.client.java.model.Partition;
import org.zalando.nakadi.client.java.model.Problem;

import com.google.common.collect.Lists;

public final class ModelFactory {

    private static Random randomizer = new Random();

    public static String randomUUID() {
        return UUID.randomUUID().toString();
    }

    public static BatchItemResponse newBatchItemResponse() {
        String detail = "detail";
        BatchItemStep step = BatchItemStep.ENRICHING;
        BatchItemPublishingStatus publishingStatus = BatchItemPublishingStatus.SUBMITTED;
        String eid = randomUUID();
        return new BatchItemResponse(eid, publishingStatus, step, detail);
    }

    public static BusinessEvent newBusinessEvent() {
        return new BusinessEventImpl(newEventMetadata());
    }

    public static EventMetadata newEventMetadata() {
        return new EventMetadata(randomUUID(), //
                "eventType", "occurredAt", //
                "receivedAt", Lists.newArrayList("parentEids"), "flowId", "partition");
    }

    public static Cursor newCursor() {
        return new Cursor(randomUUID(), randomUUID());
    }

    public static SimpleEvent newSimpleEvent() {
        return new SimpleEvent(randomUUID());
    }

    public static <T extends Event> DataChangeEvent<T> newDataChangeEvent(T event) {
        return new DataChangeEvent<T>(event, randomUUID(), DataOperation.CREATE, newEventMetadata());
    }

    public static <T extends Event> EventStreamBatch<T> newEventStreamBatch(List<T> events, Cursor cursor) {
        return new EventStreamBatch<T>(cursor, events);
    }

    public static EventTypeSchema newEventTypeSchema() {
        String schema = "{'propeties':{'key':'value'}}".replaceAll("'", "\"");
        return new EventTypeSchema(SchemaType.JSON, schema);
    }

    public static Integer randomInt() {
        return randomizer.nextInt();
    }

    public static EventTypeStatistics newEventTypeStatistics() {
        return new EventTypeStatistics(randomInt(), randomInt(), randomInt(), randomInt());
    }

    public static Metrics newMetrics() {
        int depth = 5;
        Map<String, Object> gauges = new HashMap<String, Object>();
        gauges.put("normalString", randomUUID());
        gauges.put("SimpleEvent", newSimpleEvent());
        gauges.put("List", randomListOfString(depth));
        gauges.put("List", randomListOfString(depth));
        gauges.put("MapOfMaps", randomMapOfString(depth));
        return new Metrics("version", gauges);
    }

    public static List<String> randomListOfString(int nrOfItems) {
        return Stream.generate(() -> randomUUID()).limit(nrOfItems).collect(Collectors.toList());

    }

    /**
     * A function that call itself to generate maps that contain other maps as values.
     * 
     * @param nrOfItems
     * @return
     */
    public static Map<String, Object> randomMapOfString(int nrOfItems) {
        return Stream.generate(() -> randomUUID()).limit(nrOfItems).collect(Collectors.toMap(i -> "nrOfItems-" + nrOfItems + "--" + i, i -> randomMapOfString(nrOfItems - 1)));
    }

    public static Partition newPartition() {

        return new Partition(randomUUID(), randomUUID(), randomUUID());
    }

    public static Problem newProblem() {

        return new Problem(randomUUID(), randomUUID(), randomInt(), randomUUID(), randomUUID());
    }

    public static EventType newEventType() {
        List<EventEnrichmentStrategy> newEnrichmentStrategy = Lists.newArrayList(EventEnrichmentStrategy.METADATA);
        return new EventType(randomUUID(), randomUUID(), EventTypeCategory.BUSINESS, //
                newEnrichmentStrategy, newPartitionStrategy(), newEventTypeSchema(), randomListOfString(12), randomListOfString(21), newEventTypeStatistics());
    }

    public static PartitionStrategy newPartitionStrategy() {
        return PartitionStrategy.values()[randomizer.nextInt(PartitionStrategy.values().length)];
    }

    public static EventEnrichmentStrategy newEventEnrichmentStrategy() {

        return EventEnrichmentStrategy.values()[randomizer.nextInt(EventEnrichmentStrategy.values().length)];
    }

    public static EventTypeCategory newEventTypeCategory() {
        return EventTypeCategory.values()[randomizer.nextInt(EventTypeCategory.values().length)];
    }

    public static ClientError newClientError() {
        return new ClientError(randomUUID(), Optional.of(randomInt()), Optional.of(new IllegalStateException()));
    }

}
